core/src/main/java/io/grpc/internal/ManagedChannelServiceConfig.java

```diff
@@ -0,0 +1,282 @@
+/*
+ * Copyright 2019 The gRPC Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.internal;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
+import static com.google.common.base.Verify.verify;
+
+import com.google.common.base.MoreObjects;
+import com.google.common.base.Objects;
+import com.google.common.base.Strings;
+import io.grpc.MethodDescriptor;
+import io.grpc.Status.Code;
+import java.util.Collections;
+import java.util.EnumSet;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.logging.Logger;
+
+/**
+ * {@link ManagedChannelServiceConfig} is a fully parsed and validated representation of service
+ * configuration data.
+ */
+final class ManagedChannelServiceConfig {
+
+  private static final Logger logger =
+      Logger.getLogger(ManagedChannelServiceConfig.class.getName());
+
+  private final Map<String, MethodInfo> serviceMethodMap;
+  private final Map<String, MethodInfo> serviceMap;
+
+  ManagedChannelServiceConfig(
+      Map<String, MethodInfo> serviceMethodMap,
+      Map<String, MethodInfo> serviceMap) {
+    this.serviceMethodMap = Collections.unmodifiableMap(new HashMap<>(serviceMethodMap));
+    this.serviceMap = Collections.unmodifiableMap(new HashMap<>(serviceMap));
+  }
+
+  static ManagedChannelServiceConfig fromServiceConfig(
+      Map<String, ?> serviceConfig,
+      boolean retryEnabled,
+      int maxRetryAttemptsLimit,
+      int maxHedgedAttemptsLimit) {
+    Map<String, MethodInfo> serviceMethodMap = new HashMap<>();
+    Map<String, MethodInfo> serviceMap = new HashMap<>();
+
+    // Try and do as much validation here before we swap out the existing configuration.  In case
+    // the input is invalid, we don't want to lose the existing configuration.
+
+    List<Map<String, ?>> methodConfigs =
+        ServiceConfigUtil.getMethodConfigFromServiceConfig(serviceConfig);
+
+    if (methodConfigs == null) {
+      // this is surprising, but possible.
+      logger.fine("no methods configs in service config");
```

Comment:
A fine-level log is hardly noticeable. If you want this to be more useful, I would pass the ChannelLogger to `fromServiceConfig()`, and log this information with `DEBUG`. It'd have a better chance of being noticed as it will show up when someone greps the channel logs.

Reply:
Removed.